web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
 	return fmt.Sprintf("Session: ID=%s, IsNew=%t, %s", session.ID, session.IsNew, out)
 }
 
-// OAuth2Login tries to authenticate against an oauth2 provider.
-func OAuth2Login(logger *zap.SugaredLogger, db database.Database) echo.HandlerFunc {
-	return func(c echo.Context) error {
-		w := c.Response()
-		r := c.Request()
+// OAuth2Login redirects user to the provider's sign in page or, if user is already signed in with provider,
+// authenticates the user in the background.
+func OAuth2Login(logger *zap.SugaredLogger, authConfig *AuthConfig, secret string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		fmt.Println("AUTH2LOGIN started with: ", r.Method) // tmp
+		if r.Method != "GET" {
+			authenticationError(logger, w, fmt.Sprintf("illegal request method: %s", r.Method))
+			return
+		}
 
-		provider, err := gothic.GetProviderName(r)
+		// Start or refresh the session.
+		// Issue: server generates a new random key to encode sessions on each restart.
+		// A session from before the restart will be detected, but cannot be decoded with
+		// the new key, resulting in an error. Instead of returning, we attempt to refresh the session first.
+		s, err := sessionStore.Get(r, SessionKey)
 		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
-		}
-		var teacher int
-		if strings.HasSuffix(provider, TeacherSuffix) {
-			teacher = 1
+			if err := sessionStore.Save(r, w, s); err != nil {
+				authenticationError(logger, w, fmt.Sprintf("failed to create new session (%s): %s", SessionKey, err))
+				return
+			}
 		}
-		logger.Debugf("Provider: %v ; Teacher: %v", provider, teacher)
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect := extractRedirectURL(r, Redirect)
-		logger.Debugf("redirect: %v", redirect)
-		// TODO: Add a random string to protect against CSRF.
-		qv.Set(State, strconv.Itoa(teacher)+redirect)
-		logger.Debugf("State: %v", strconv.Itoa(teacher)+redirect)
-		r.URL.RawQuery = qv.Encode()
-		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
+		// Get provider name.
+		provider := GetProviderName(r.URL.Path, 2)
```

Comment:
Can you please add a test for the `GetProviderName()` function with all the possible `URL.Path`s that it might receive.

I think it should be refactored...

I think it should return an error instead of an empty string. The error could then be passed on to `authError`. 

Note that there are two uses of the function with different `index` input, which is weird to me. What type of `URL.Path` input are expected here? And why do we need to pass the `index` input anyway?

Reply:
On second thought, why not hide this (provider lookup from `URL.Path`) in a method on `AuthConfig` that takes the `r` as input (similar to gothic).

Reply:
I wonder if we'd rather have different endpoints for different providers and refactor the common methods? I feel that reading anything from URLs is just wrong... 

This way we could have a constant `/auth/github` endpoint just for GitHub interactions, and add more if needed, like `/auth/gitlab` and so on.

Reply:
I agree that examining the URLs is the wrong approach. I believe [my comment](https://github.com/quickfeed/quickfeed/pull/692/files#r921802570) mentioned above, resolves this problem, obviating the need for the `GetProviderName()` method altogether.